sway-core/src/semantic_analysis/cei_pattern_analysis.rs

```diff
@@ -0,0 +1,258 @@
+        },
+        Tuple { fields: exprs }
+        | Array { contents: exprs } => {
+            effects_of_expressions(exprs)
+        },
+        StructExpression {fields, ..} => {
+            effects_of_struct_expressions(fields)
+        },
+        CodeBlock(codeblock) => {
+            effects_of_codeblock(codeblock)
+        },
+        IfExp { condition, then, r#else } => {
+            let mut effs = effects_of_expression(condition);
+            effs.extend(effects_of_expression(then));
+            let else_effs =
+                match r#else {
+                    Some(expr) => effects_of_expression(expr),
+                    None => HashSet::new(),
+                };
+            effs.extend(else_effs);
+            effs
+        },
+        StructFieldAccess { prefix: expr, ..}
+        | TupleElemAccess { prefix: expr, ..}
+        | EnumTag { exp: expr }
+        | UnsafeDowncast {exp: expr, ..}
+        | Return(expr) => {
+            effects_of_expression(expr)
+        },
+        EnumInstantiation {contents, ..} => {
+            match contents {
+                Some(expr) => effects_of_expression(expr),
+                None => HashSet::new(),
+            }
+        },
+        AbiCast {address, ..} => {
+            effects_of_expression(address)
+        },
+        IntrinsicFunction(intr_fn) => {
+            effects_of_expressions(&intr_fn.arguments)
+            .union(&effects_of_intrinsic(&intr_fn.kind))
+            .cloned()
+            .collect()
+        },
+        WhileLoop { condition, body } => {
+            effects_of_expression(condition)
+            .union(&effects_of_codeblock(body))
+            .cloned()
+            .collect()
+        },
+        FunctionApplication {function_decl_id,
+                             arguments,
+                             selector, ..} => {
+
+            use crate::declaration_engine::de_get_function;
+            let fn_body = de_get_function(function_decl_id.clone(), &expr.span).unwrap().body;
+            let mut effs = effects_of_codeblock(&fn_body);
+            let args_effs = map_hashsets_union(arguments, |e| effects_of_expression(&e.1));
+            effs.extend(args_effs);
+            if selector.is_some() {
+                // external contract call (a.k.a. interaction)
+                effs.extend(HashSet::from([Effect::Interaction]))
+            };
+            effs
+        },
+        // TODO: process assembly blocks
+        AsmExpression {
+            registers: _,
+            body: _,
+            whole_block_span: _,
+            ..
+        } => {
+            // temporary solution, will remove after this is finished
+            HashSet::new()
+        },
+    }
+}
+
+fn effects_of_intrinsic(intr: &sway_ast::Intrinsic) -> HashSet<Effect> {
+    use sway_ast::Intrinsic::*;
+    match intr {
+        StateStoreWord | StateStoreQuad => HashSet::from([Effect::StorageWrite]),
+        // TODO: figure out the effect of __revert
```

Comment:
I think a revert (or a return) between a contract call and a storage write should not result in a reentrancy warning. That being said, we should probably be warning on dead code in that case anyways?